### PR TITLE
feature/interactive-history-stats

### PR DESCRIPTION
**Pull request by haritonasty**
**Summary**
1) Reworked history widget for watchlist
2) rewrote some gql queries to new api (get history by watchlist id instead array of slugs)

[**Video**](https://www.loom.com/share/99e301a120dc4535bb1056abedbfef7c)

**Screenshots**
image

### CODE REVIEW

File: src/components/WatchlistHistory/GetWatchlistHistory.js [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+import { TOTAL_TYPES } from './constants'
+import {
+  CATEGORY_HISTORY_QUERY,
+  PROJECTS_HISTORY_QUERY,
+  WATCHLIST_HISTORY_QUERY
+} from './WatchlistHistoryGQL'
+import WatchlistHistoryWidget from './WatchlistHistoryWidget'
+import { filterEmptyStats } from './utils'
+
+const getMarketcapQuery = ({ type, projects, range, byFunction }) => {
```
Comment (Partysun) [inline src/components/WatchlistHistory/GetWatchlistHistory.js]: byFunction used for true/false.
Is it ok name for this? @haritonasty

File: src/components/WatchlistHistory/GetWatchlistHistory.js [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+  return byFunction
+    ? graphql(PROJECTS_HISTORY_QUERY, {
+      props: ({
+        data: { projectsListHistoryStats = [], loading: isLoading }
+      }) => ({
+        historyPrice: filterEmptyStats(projectsListHistoryStats),
+        isLoading
+      }),
+      options: () => ({
```
Comment (Partysun) [inline src/components/WatchlistHistory/GetWatchlistHistory.js]: options before props. We used it often. And you use it in the next graphql func.

File: src/components/WatchlistHistory/GetWatchlistHistory.js [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    const HistoryQuery = resultQuery(WatchlistHistoryWidget)
+    return (
+      <HistoryQuery
+        {...rest}
+        type={TOTAL_TYPES[type] ? 'Total' : 'Watchlist'}
+        interval={range.value}
+        combinedPeriod={range.interval}
+      />
+    )
+  } else return null
```
Comment (Partysun) [inline src/components/WatchlistHistory/GetWatchlistHistory.js]: You can return null without else

File: src/components/WatchlistHistory/GetWatchlistHistory.js [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+}) => {
+  if (range) {
+    const resultQuery = getMarketcapQuery({ type, projects, range, byFunction })
+    const HistoryQuery = resultQuery(WatchlistHistoryWidget)
+    return (
+      <HistoryQuery
+        {...rest}
+        type={TOTAL_TYPES[type] ? 'Total' : 'Watchlist'}
+        interval={range.value}
+        combinedPeriod={range.interval}
```
Comment (Partysun) [inline src/components/WatchlistHistory/GetWatchlistHistory.js]: interval = range.value
combinedPeriod = range.interval
Feel like interval should be interval, but ok.)

File: src/components/WatchlistHistory/WatchlistHistoryTemplate.js [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+              padding
+            >{`Calculated as average value for the last ${combinedPeriod}. You can change period by pressing button`}</Panel>
+          </Tooltip>
+        </div>
+        <div className={styles.bottom}>
+          <span className={styles.value}>$ {value}</span>
+          {!isLoading && (
+            <PercentChanges changes={change} className={styles.change} />
+          )}
+          {isLoading && <Loader className={styles.loader} />}
```
Comment (Partysun) [inline src/components/WatchlistHistory/WatchlistHistoryTemplate.js]: First isLoading as a positive condition.

File: src/components/WatchlistHistory/constants.js [hunk 6]
```diff
@@ -0,0 +1,21 @@
+import { DAY, getTimeIntervalFromToday, MONTH } from '../../utils/dates'
+
+export const TOTAL_TYPES = {
+  all: 'TOTAL_MARKET',
+  erc20: 'TOTAL_ERC20'
+}
+
+const RANGES_METHOD = new Map()
```
Comment (Partysun) [inline src/components/WatchlistHistory/constants.js]: Do we have this in our source in other place?

File: src/components/WatchlistHistory/GetWatchlistHistory.js (hunk 1 above)
Comment (haritonasty) [reply src/components/WatchlistHistory/GetWatchlistHistory.js]: isByFunction better?)

File: src/components/WatchlistHistory/constants.js (hunk 6 above)
Comment (DmitriVanGuard) [reply src/components/WatchlistHistory/constants.js]: @Partysun @haritonasty [`getIntervalByTimeRange`](https://github.com/santiment/app/blob/3a6d4a9efbc088fdd5086dd76c6a77e15cb010c1/src/utils/dates.js#L280) is our parser for such cases. Returns `{from: Date, to: Date}`

**Pull request merged**